File: libgraffiti/src/css/values/box_shadow.rs
use super::super::parsing::{sym, Parsable, Parser};
use super::{Color, Px};
use std::fmt;

#[derive(Debug, Clone, Copy, PartialEq)]
pub struct BoxShadow {
    pub offset: (Px, Px),
    pub blur: Px,
    pub spread: Px,
    pub color: Color,
}

impl fmt::Display for BoxShadow {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(
            f,
            "{} {} {} {} {}",
            self.offset.0, self.offset.1, self.blur, self.spread, self.color
        )
    }
}

impl Parsable for BoxShadow {
    fn parser<'a>() -> Parser<'a, Self> {
        let offset = Px::parser() - sym(" ") + Px::parser();
        let blur = Px::parser();
        let spread = (Px::parser() - sym(" ")).opt().map(|d| d.unwrap_or(Px(0.)));
        let shadow = offset - sym(" ") + blur - sym(" ") + spread + Color::parser();

        shadow.map(|(((offset, blur), spread), color)| Self {
            offset,
            blur,
            spread,
            color,
        })
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_shadow() {
        assert_eq!(
            BoxShadow::parse("1px 1px 1px 1px #000"),
            Ok(BoxShadow {
                offset: (Px(1.), Px(1.)),
                blur: Px(1.),
                spread: Px(1.),
                color: Color::BLACK
            })
        );

        assert_eq!(
            BoxShadow::parse("0 0 10px #000"),
            Ok(BoxShadow {
                offset: (Px(0.), Px(0.)),
                blur: Px(10.),
                spread: Px(0.),
                color: Color::BLACK
            })
        );
    }
}

File: libgraffiti/src/css/values/BoxShadow.zig
const std = @import("std");
const Parser = @import("../parser.zig").Parser;
const Px = @import("./Px.zig").Px;
const Color = @import("./Color.zig").Color;

pub const BoxShadow = struct {
    offset_x: Px,
    offset_y: Px,
    blur: Px,
    spread: Px,
    color: Color,

    const Self = @This();

    pub fn parse(parser: *Parser) !Self {
        return Self{
            .offset_x = try parser.parse(Px),
            .offset_y = try parser.parse(Px),
            .blur = try parser.parse(Px),
            // TODO: opt
            .spread = try parser.parse(Px),
            .color = try parser.parse(Color),
        };
    }
};

fn expectBoxShadow(input: []const u8, expected: BoxShadow) !void {
    try std.testing.expectEqual(expected, try Parser.init(std.testing.allocator, input).parse(BoxShadow));
}

test "Shadow.parse()" {
    try expectBoxShadow(
        "1px 1px 1px 1px #000",
        BoxShadow{
            .offset_x = .{ .px = 1 },
            .offset_y = .{ .px = 1 },
            .blur = .{ .px = 1 },
            .spread = .{ .px = 1 },
            .color = Color.BLACK,
        },
    );

    // try expectBoxShadow(
    //     "0 0 10px #000",
    //     BoxShadow{
    //         .offset_x = .{ .px = 0 },
    //         .offset_y = .{ .px = 0 },
    //         .blur = .{ .px = 10 },
    //         .spread = .{ .px = 0 },
    //         .color = Color.BLACK,
    //     },
    // );
}
